src/bsoncxx/test/bson_validate.cpp

```diff
@@ -0,0 +1,135 @@
+#include "catch.hpp"
+
+#include <bsoncxx/validate.hpp>
+
+#include <array>
+
+#include <bsoncxx/builder/stream/document.hpp>
+#include <bsoncxx/builder/stream/helpers.hpp>
+
+using namespace bsoncxx;
+
+// These are basic sanity tests to make sure the API works.
+// Libbson has a much more complete suite of tests that covers various corrupt BSON
+// inputs.
+
+namespace {
+    // polymorphic lambdas would be nice here.
+    template<typename T> 
+    bool is_engaged(const stdx::optional<T>& opt) {
+        return opt != stdx::nullopt;
```

Comment:
Can you just return static_cast<bool>(opt) to explictly invoke the explicit conversion to bool?